### PR TITLE
Store Computed Colors on `nodes` and `links`

### LINKED ISSUE #123: Option to Store Computed Colors on `nodes` and `links`
## Elevator Pitch

* Like with `NodeSelection` and `LinkSelection`, add an option to `ContinuousColor` and `OrdinalColor` to store the value in a column.

## Motivation

* We want to be able to color `links` based on the color of their `source` or `target`, but when the `nodes` are colored using `ContinuousColor`, it is not possible to reference a `node`'s color.
* This is motivated from the concept of groups in a graph, where we not only want to color the `nodes` based on their group assignment, but also the `links` that connect between two `nodes` in the same group.

## Design Ideas

* Add a `column_name` trait to `ContinuousColor` and `OrdinalColor` and store the computed color under that name,
* As with `NodeSelection` and `LinkSelection`, the `default_value = None` and in which case, nothing is stored.
* We may want to check that the user does not specify `__indexColor` as that seems to be a reserved property in `forcegraph`.

### PR DESCRIPTION
**Pull request by sanbales**
## Checklist

- [x] the PR target is correct
  - `main` for releases, docs-only fixes and post-release hot-fixes
  - `dev` for everything else
- [x] ran `doit lint` locally

## References

* fixes #123 

## Code changes

- [x] Added `column_name` to `ContinuousColor` and `OrdinalColor`.
- [x] tests
  - [x] unit test
  - [x] robot test
- [x] docs
  - [x] Update DOCSTRINGs
  - [x] update `Scales.ipynb` to demonstrate the new functionality
- [x] changelog

## User-facing changes

| note | shots |
|-|-|
| reference the computed color of `nodes` when coloring `links` | image image
| screen capture from `Scales.ipynb` | image

### CODE REVIEW

File: js/widgets/behaviors/base.ts [hunk 1]
```diff
@@ -243,7 +243,7 @@ export class NunjucksModel extends DynamicModel {
     }
 
     function handler(opts: any) {
-      const value = coercer(tmpl.render(opts));
+      const value = coercer(tmpl.render(opts).trim());
```
Comment (sanbales) [inline js/widgets/behaviors/base.ts:246]: @nrbgt : not sure if this is will impact performance, but if it does not, it helps read the rendered template when troubleshooting

File: src/ipyforcegraph/behaviors/scales.py [hunk 2]
```diff
@@ -70,8 +89,10 @@ class Scheme(enum.Enum):
         allow_none=True,
     ).tag(sync=True)
 
-    domain: Tuple[float, float] = W.TypedTuple(
-        (0.0, 1.0),
+    domain: Tuple[float, float] = T.Tuple(
+        T.Float(),
```
Comment (sanbales) [inline src/ipyforcegraph/behaviors/scales.py:93]: Added this more specific trait definitions, but happy to go back to `TypedTuple` if you think that's better.

File: examples/Scales.ipynb
Comment (sanbales) [inline examples/Scales.ipynb:1]: @nrbgt : I'm not sure why the colors are not displaying. I check the rendering of the nunjucks template and it seems to be getting the right values, e.g., for the links colored by nodes in the console I can see that the `Nunjucks` template evaluates to `"gray"` for links that have `source.value != target.value` and a color for the ones that match.

Comment (sanbales) [reply examples/Scales.ipynb:1]: Resolved, looks like the issue was comparing float values.  Also might have been due to the unnecessary column check removed in [3ec2dc0](https://github.com/jupyrdf/ipyforcegraph/pull/124/commits/3ec2dc0ace7de5ffa82eed22c8e295a78c879296)

File: js/widgets/behaviors/base.ts (hunk 1 above)
Comment (sanbales) [reply js/widgets/behaviors/base.ts:246]: Resolved in [df42d1f](https://github.com/jupyrdf/ipyforcegraph/pull/124/commits/df42d1f49ef6a055a40aad25320f20fbe6a64e95)

**Comment by nrbgt** [comment]
With the changes updated it seems to do the thing, though looking at the title, this is not quite accurate.

Would there be a broader value in changing the locus of this to the `Behavior` level, where _every_ behavior/facet can update a computed value on the graph?

**Comment by sanbales** [comment]
> With the changes updated it seems to do the thing, though looking at the title, this is not quite accurate.
> 
> Would there be a broader value in changing the locus of this to the `Behavior` level, where _every_ behavior/facet can update a computed column on the source? E.g. `save_as_column_name`.
> 
> Of course, this would only be available with the semi-flaky recorder (or on `Nunjucks`-based facets/behaviors), but would likely be more consistent.
> 
> Another approach, rather than changing all those APIs, would be to create a `CaptureAs` which could wrap anything else.
> 
> ```python
> ns = NodeShapes(color=Capture("some_column", OrdinalColorScheme(...)))
> ```

I like the idea of just having a property (e.g., `save_as_column_name`) because it seems easier for people to discover the functionality, but I'm not sure how we would do different facets of the behavior, e.g., `color` and `width` for `LinkShapes`.  That's where the `Capture` would be cleaner, and at first I was concerned about adding complexity to the API but the more I think about it, it seems that that would be the way to go.  Would we leave the `NodeSelection`/`LinkSelection` `column_name` as is, or would we modify it to work in this new way?

**Comment by nrbgt** [comment]
> just having a property

right, but we'd have to add it almost everywhere instead of one or two checks.

> discover the functionality

Not so worried about that, as it's a sub-optimal performance/consistency pattern which we haven't even really discovered all the issues with. indeed, perhaps we should have added a `ReplaceCssVariables` as this kind of functional approach... when we were debugging this PR the other day, I noted where it was being called twice on something, which certainly isn't great. _Maybe_ it's not such a big deal to refactor out (though would be a breaking change, e.g. `0.4.0`) but it probably secretly _was_ a breaking change if you were emitting really fat HTML with `style` on tooltips.

>  Would we leave the NodeSelection/LinkSelection column_name as is, or would we modify it to work in this new way?

To my recollection, they work substantially differently (event based with internal state) than the purely declarative ones, so they would not need to change. It could however point to a pattern where there was a `(Node|Link)Event` superclass for #51, but we'd probably leave  the existing`*Selection` API as-is as to not introduce a breaking change.

**Comment by sanbales** [comment]
> right, but we'd have to add it almost everywhere instead of one or two checks.

Yeah, good point.

> discover the functionality

Concur, was more concerned at first, but the more I think about it, I agree with with you.

> Would we leave the NodeSelection/LinkSelection column_name as is, or would we modify it to work in this new way?

I think you are right, they are substantially different, was just asking in case there was some commonality because of the whole add data to the nodes thing.

>  perhaps we should have added a `ReplaceCssVariables`

Refactoring the `ReplaceCssVariables` with this functional approach does sound like a good improvement, we could then have a new pattern to handle these types of operations that seems (in my limited experience/understanding) to be more scalable.  Also fully concur we would need to bump to `0.4.0`.

**Comment by sanbales** [comment]
Supplanted by #125

**Pull request closed without merging**